File: src/klaude_code/server/system_context.py
"""On-demand system prompt + tool catalogue for the viewer's SYSTEM inspector.

Neither the system prompt nor the tool schemas belong in a status frame that
refreshes on every model change -- together they are tens of kilobytes -- so the
viewer asks for them once, when a reader opens the SYSTEM row.

Two sources, and the answer says which one it is:

- ``live``: the loaded agent's own profile, i.e. the exact ``system`` / ``tools``
  the next step would send.
- ``rebuilt``: for a cold session, the same builders run again off the session
  meta (model name, work_dir, agent type). Prompt assembly reads prompt files
  and probes the work_dir; it never constructs an LLM client, touches the
  network, or writes anything. It reflects TODAY's prompt files and tool set,
  not what the session actually sent, so readers must label it as rebuilt.

Model config is reported through an explicit allowlist of display fields rather
than a model dump: no credential can reach the response by being added upstream.
"""

from __future__ import annotations

import time
from pathlib import Path
from typing import Any, Final

from klaude_code.agent.agent_profile import AgentProfile, load_agent_tools
from klaude_code.agent.system_prompt import load_system_prompt
from klaude_code.protocol import llm_param
from klaude_code.tool.core.registry import get_tool_schemas

CACHE_TTL_SECONDS: Final = 30.0

# session_id -> (expires_at_monotonic, payload)
_CACHE: dict[str, tuple[float, dict[str, Any]]] = {}

# Vanilla sessions strip the profile down to these; mirrors
# ``agent/agent_profile.py:VanillaModelProfileProvider``.
_VANILLA_SYSTEM_PROMPT: Final = "You're an agent running in user's terminal"
_VANILLA_TOOLS: Final = ["Bash", "Edit", "Write", "Read"]


def clear_cache() -> None:
    _CACHE.clear()


def _tool_payload(schemas: list[llm_param.ToolSchema]) -> list[dict[str, Any]]:
    return [{"name": tool.name, "description": tool.description, "parameters": tool.parameters} for tool in schemas]


def _live_model_payload(config: llm_param.LLMConfigParameter, model_config_name: str | None) -> dict[str, Any]:
    """Allowlisted display fields. Never widen this to a model dump."""
    thinking = config.thinking
    return {
        "provider": config.provider_name or None,
        "protocol": config.protocol.value,
        "model": config.model_id,
        "model_config_name": model_config_name,
        "effort": config.effective_effort,
        "max_tokens": config.max_tokens,
        "context_limit": config.context_limit,
        "temperature": config.temperature,
        "verbosity": config.verbosity,
        "thinking": thinking.model_dump(mode="json", exclude_none=True) if thinking is not None else None,
        "fast_mode": config.fast_mode,
        "cache_retention": config.cache_retention,
        "supports_vision": config.supports_vision,
    }


def _cold_model_payload(meta: dict[str, Any]) -> dict[str, Any]:
    """What meta.json remembers about the model. No config lookup, no secrets."""
    thinking = meta.get("model_thinking")
    effort = meta.get("model_effort")
    if not isinstance(effort, str):
        effort = None
    return {
        "provider": None,
        "protocol": None,
        "model": meta.get("model_name") if isinstance(meta.get("model_name"), str) else None,
        "model_config_name": meta.get("model_config_name") if isinstance(meta.get("model_config_name"), str) else None,
        "effort": effort,
        "max_tokens": None,
        "context_limit": None,
        "temperature": None,
        "verbosity": None,
        "thinking": thinking if isinstance(thinking, dict) else None,
        "fast_mode": None,
        "cache_retention": None,
        "supports_vision": None,
    }


def live_payload(profile: AgentProfile, *, model_config_name: str | None) -> dict[str, Any]:
    return {
        "available": True,
        "source": "live",
        "system_prompt": profile.system_prompt,
        "tools": _tool_payload(profile.tools),
        "model": _live_model_payload(profile.llm_client.get_llm_config(), model_config_name),
        "reason": None,
    }


def rebuilt_payload(meta: dict[str, Any], work_dir: Path) -> dict[str, Any]:
    """Re-run the prompt/tool builders off session meta. Read-only."""
    model_name = meta.get("model_name")
    if not isinstance(model_name, str) or not model_name:
        config_name = meta.get("model_config_name")
        model_name = config_name if isinstance(config_name, str) else ""
    if bool(meta.get("vanilla", False)):
        return {
            "available": True,
            "source": "rebuilt",
            "system_prompt": _VANILLA_SYSTEM_PROMPT,
            "tools": _tool_payload(get_tool_schemas(_VANILLA_TOOLS)),
            "model": _cold_model_payload(meta),
            "reason": None,
        }
    agent_type = meta.get("agent_type")
    sub_agent_type = agent_type if isinstance(agent_type, str) and agent_type else None
    try:
        # supports_vision is a per-model config flag the meta does not keep; the
        # only effect of the default is whether LookAt is appended.
        tools = load_agent_tools(model_name, sub_agent_type, supports_vision=True)
        system_prompt = load_system_prompt(model_name, sub_agent_type, available_tools=tools, work_dir=work_dir)
    except (KeyError, OSError, ValueError) as exc:
        return unavailable(f"rebuild failed: {exc.__class__.__name__}: {exc}")
    return {
        "available": True,
        "source": "rebuilt",
        "system_prompt": system_prompt,
        "tools": _tool_payload(tools),
        "model": _cold_model_payload(meta),
        "reason": None,
    }


def unavailable(reason: str) -> dict[str, Any]:
    return {
        "available": False,
        "source": None,
        "system_prompt": None,
        "tools": None,
        "model": None,
        "reason": reason,
    }


def cached(session_id: str) -> dict[str, Any] | None:
    hit = _CACHE.get(session_id)
    if hit is None:
        return None
    expires_at, payload = hit
    if expires_at <= time.monotonic():
        _CACHE.pop(session_id, None)
        return None
    return payload


def store(session_id: str, payload: dict[str, Any]) -> dict[str, Any]:
    _CACHE[session_id] = (time.monotonic() + CACHE_TTL_SECONDS, payload)
    return payload


__all__ = [
    "CACHE_TTL_SECONDS",
    "cached",
    "clear_cache",
    "live_payload",
    "rebuilt_payload",
    "store",
    "unavailable",
]

File: web/src/adapter/system-context.ts
/**
 * `GET /api/web/sessions/{id}/system-context` -> the SYSTEM row's prompt snapshot.
 *
 * klaude never persists the system prompt or the tool catalogue (decision #16 /
 * plan "记录检查器"), so the viewer asks for them once per page load and the
 * adapter hangs the answer on the first request of the loaded window. That is
 * what makes `layout.ts` emit its `system` record ("Initial System Prompt")
 * with the System Prompt / Tools tabs.
 *
 * A `rebuilt` answer is not what the session sent: the server re-ran today's
 * prompt builders off the session meta. The snapshot therefore carries a
 * `caveat` the inspector prints, so nobody reads a cold session's SYSTEM row as
 * a recording.
 */

import type { ConversationPromptSnapshot, AssistantRequestConfig, ToolSchema } from '../contract/index.ts'
import type { SystemContext, SystemContextModel } from './wire.ts'

/** Locale keys this module asks the caller's translator for. */
export const REBUILT_CAVEAT_KEY = 'klaude.systemContext.rebuilt'

function text(value: string | null | undefined): string | undefined {
  return typeof value === 'string' && value !== '' ? value : undefined
}

function finite(value: number | null | undefined): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined
}

/**
 * Model knobs -> the inspector's request config.
 *
 * Only the fields the vendored `AssistantRequestConfig` declares are carried;
 * the rest of the allowlist (`verbosity`, `cache_retention`, `fast_mode`,
 * `context_limit`, `supports_vision`) has no upstream slot and is dropped
 * rather than smuggled in under a wrong name.
 * @param model - The endpoint's `model` object, when it sent one.
 * @returns The config; provider/model fall back to empty strings.
 */
export function promptConfig(model: SystemContextModel | null | undefined): AssistantRequestConfig {
  const thinking = model?.thinking
  const effort = text(model?.effort)
  const temperature = finite(model?.temperature)
  const maxTokens = finite(model?.max_tokens)
  return {
    provider: text(model?.provider) ?? '',
    model: text(model?.model) ?? text(model?.model_config_name) ?? '',
    ...(effort === undefined ? {} : { reasoningEffort: effort }),
    ...(temperature === undefined ? {} : { temperature }),
    ...(maxTokens === undefined ? {} : { maxTokens }),
    ...(thinking === undefined || thinking === null
      ? {}
      : { thinking: JSON.stringify(thinking) }),
  }
}

/**
 * The tool catalogue, in the shape the Tools tab and the Schema tab read.
 * @param context - The endpoint payload, when it has been fetched.
 * @returns Tool schemas in server order; empty when none were reported.
 */
export function promptTools(context: SystemContext | undefined): readonly ToolSchema[] {
  const tools = context?.tools
  if (tools === undefined || tools === null) return []
  return tools.map(tool => ({
    name: tool.name,
    description: tool.description,
    parameters: tool.parameters,
  }))
}

/**
 * Build the SYSTEM row's prompt snapshot.
 * @param context - The endpoint payload; `undefined` before it is fetched.
 * @param translate - Locale seat for the `rebuilt` caveat.
 * @returns The snapshot, or undefined when the endpoint reported
 *   `available: false` (then there is no SYSTEM row at all).
 */
export function promptSnapshot(
  context: SystemContext | undefined,
  translate?: (key: string) => string,
): ConversationPromptSnapshot | undefined {
  if (context === undefined || !context.available) return undefined
  const caveat = context.source === 'rebuilt'
    ? translate?.(REBUILT_CAVEAT_KEY) ?? REBUILT_CAVEAT_KEY
    : undefined
  return {
    config: promptConfig(context.model),
    system: context.system_prompt ?? '',
    tools: promptTools(context),
    ...(caveat === undefined ? {} : { caveat }),
  }
}

/**
 * Index the catalogue by tool name so a call id can be resolved to its schema.
 * @param context - The endpoint payload, when it has been fetched.
 * @returns name -> schema; empty when the catalogue is unavailable.
 */
export function toolSchemasByName(
  context: SystemContext | undefined,
): ReadonlyMap<string, ToolSchema> {
  const byName = new Map<string, ToolSchema>()
  if (context === undefined || !context.available) return byName
  for (const tool of promptTools(context)) byName.set(tool.name, tool)
  return byName
}
